File: R/gompertz.R
##' Gompertz model with log-normal observations.
##'
##' \code{gompertz()} constructs a \sQuote{pomp} object encoding a stochastic Gompertz population model with log-normal measurement error.
##'
##' The state process is \eqn{X_{t+1} = K^{1-S} X_{t}^S
##' \epsilon_{t}}{X[t+1]=K^(1-S) X[t]^S eps[t]}, where \eqn{S=e^{-r}}{S=e^{-r}}
##' and the \eqn{\epsilon_t}{eps[t]} are i.i.d. lognormal random deviates with
##' variance \eqn{\sigma^2}{sigma^2}.  The observed variables \eqn{Y_t} are
##' distributed as
##' \eqn{\mathrm{lognormal}(\log{X_t},\tau)}{lognormal(log(X[t]),tau)}.
##' Parameters include the per-capita growth rate \eqn{r}, the carrying
##' capacity \eqn{K}, the process noise s.d. \eqn{\sigma}{sigma}, the
##' measurement error s.d. \eqn{\tau}{tau}, and the initial condition
##' \eqn{X_0}{X[0]}.  The \sQuote{pomp} object includes parameter
##' transformations that log-transform the parameters for estimation purposes.
##'
##' @name gompertz
##' @docType data
##' @keywords models pomp_datasets
##' @include simulate.R
##' @family pomp_examples
##'
##' @return
##' A \sQuote{pomp} object with simulated data.
##'
##' @examples
##'
##' plot(gompertz())
##' plot(gompertz(K=2,r=0.01))
##'
NULL

##' @rdname gompertz
##'
##' @param r growth rate
##' @param K carrying capacity
##' @param sigma process noise intensity
##' @param tau measurement error s.d.
##' @param X_0 value of the latent state variable \code{X} at the zero time
##' @param t0 zero time
##' @param times observation times
##'
##' @export
gompertz <- function (K = 1, r = 0.1, sigma = 0.1, tau = 0.1, X_0 = 1,
  times = 1:100, t0 = 0)
{

  simulate(
    times=times, t0=t0,
    params=c(K=K,r=r,sigma=sigma,tau=tau,X_0=X_0),
    partrans=parameter_trans(
      toEst="_gompertz_to_trans",
      fromEst="_gompertz_from_trans"
    ),
    rprocess=discrete_time(
      step.fun="_gompertz_simulator"
    ),
    rmeasure="_gompertz_normal_rmeasure",
    dmeasure="_gompertz_normal_dmeasure",
    skeleton=map("_gompertz_skeleton",delta.t=1),
    PACKAGE="pomp",
    paramnames=c("r","K","sigma","tau","X_0"),
    obsnames=c("Y"),
    statenames=c("X"),
    seed=299438676L
  )

}

File: src/gompertz.c
// dear emacs, please treat this as -*- C++ -*-

#include <Rmath.h>

#include "pomp.h" // in R, do 'system.file("include/pomp.h",package="pomp")' to find this header file

// define some macros to make the code easier to read
#define R       (p[parindex[0]]) // growth rate
#define K       (p[parindex[1]]) // carrying capacity
#define SIGMA   (p[parindex[2]]) // process noise level
#define TAU     (p[parindex[3]]) // measurement noise level

#define Y           (y[obsindex[0]])   // observed population size
#define X           (x[stateindex[0]]) // actual population size
#define XPRIME      (f[stateindex[0]]) // new population size (for skeleton function only)

// normal measurement model density
void _gompertz_normal_dmeasure (double *lik, double *y, double *x, double *p, int give_log,
  int *obsindex, int *stateindex, int *parindex, int *covindex,
  double *covars, double t) {
  *lik = dlnorm(Y,log(X),TAU,give_log);
}

// normal measurement model simulator
void _gompertz_normal_rmeasure (double *y, double *x, double *p,
  int *obsindex, int *stateindex, int *parindex, int *covindex,
  double *covars, double t) {
  Y = rlnorm(log(X),TAU);
}

// stochastic Gompertz model with log-normal process noise
void _gompertz_simulator (double *x, const double *p,
  const int *stateindex, const int *parindex, const int *covindex,
  int covdim, const double *covar, double t, double dt)
{
  double S = exp(-R*dt);
  double eps = (SIGMA > 0.0) ? exp(rnorm(0,SIGMA)) : 1.0;
  X = pow(K,(1-S))*pow(X,S)*eps; // note X is over-written by this line
}

// the deterministic skeleton
void _gompertz_skeleton (double *f, double *x, const double *p,
  const int *stateindex, const int *parindex, const int *covindex,
  const double *covar, double t)
{
  double dt = 1.0;
  double S = exp(-R*dt);
  XPRIME = pow(K,(1-S))*pow(X,S); // X is not over-written in the skeleton function
}

#undef R
#undef K
#undef SIGMA
#undef TAU
#undef Y
#undef X
#undef XPRIME

#define r		(__p[__parindex[0]])
#define K		(__p[__parindex[1]])
#define sigma		(__p[__parindex[2]])
#define tau		(__p[__parindex[3]])
#define X_0		(__p[__parindex[4]])
#define T_r		(__pt[__parindex[0]])
#define T_K		(__pt[__parindex[1]])
#define T_sigma		(__pt[__parindex[2]])
#define T_tau		(__pt[__parindex[3]])
#define T_X_0		(__pt[__parindex[4]])

void _gompertz_to_trans (double *__pt, const double *__p, const int *__parindex)
{
  T_r = log(r);
  T_K = log(K);
  T_sigma = log(sigma);
  T_tau = log(tau);
  T_X_0 = log(X_0);
}

void _gompertz_from_trans (double *__p, const double *__pt, const int *__parindex)
{
  r = exp(T_r);
  K = exp(T_K);
  sigma = exp(T_sigma);
  tau = exp(T_tau);
  X_0 = exp(T_X_0);
}

#undef r
#undef K
#undef sigma
#undef tau
#undef X_0
#undef T_r
#undef T_K
#undef T_sigma
#undef T_tau
#undef T_X_0
